tools/autograd/derivatives.yaml

```diff
@@ -649,27 +670,33 @@
   # This is because source is not broadcastable to index, as source.dim() < index.dim()
   source: "source.dim() > 0 ? grad.index_select(dim, index).expand_as(source) : grad.index_select(dim, index.squeeze(0))"
   index: non_differentiable
+  result: self_t.index_copy_(dim, index, source_t)
 
 - name: index_fill_.int_Scalar(Tensor(a!) self, int dim, Tensor index, Scalar value) -> Tensor(a!)
   self: grad.clone().index_fill_(dim, index, 0)
   index: non_differentiable
+  result: self_t.index_fill_(dim, index, 0)
 
 - name: index_fill_.int_Tensor(Tensor(a!) self, int dim, Tensor index, Tensor value) -> Tensor(a!)
   self: grad.clone().index_fill_(dim, index, 0)
   value: grad.index_select(dim, std::get<0>(at::_unique(index, /*sorted=*/false))).sum()
   index: non_differentiable
+  result: self_t.index_fill_(dim, index, value_t)
 
 - name: index_put_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor(a!)
   self: "accumulate ? grad : grad.clone().index_put_(indices, zeros_like(values), false)"
   values: grad.index(indices)
+  result: self_t.index_put_(indices, values_t, accumulate)
 
 - name: _index_put_impl_(Tensor(a!) self, Tensor?[] indices, Tensor values, bool accumulate=False, bool unsafe=False) -> Tensor(a!)
   self: "accumulate ? grad : grad.clone().index_put_(indices, zeros_like(values), false)"
   values: grad.index(indices)
+  result: at::_index_put_impl_(self_t, indices, values_t, accumulate, unsafe)
```

Comment:
Why are index_put_ and _index_put_impl_ different?

Reply:
Not sure to be honest... It does look weird. I'm not even sure why the `impl` version exists.. Is that ok if I file this as a follow up task?

Reply:
Yeah, feel free to handle this in a follow-up task. The formula looks correct to me, so I don't think anything (aside from test coverage) could go wrong

Reply:
After looking into more details, I think the reason for this is that an extra argument was needed (unsafe) but to avoid exposing it externally, the impl version was added.